File: src/collision.js
import { referenceDistance, fallRecovery } from "./attributes.js";
import { opponentDistance } from "./physical-pose.js";
const unit = 22.4 / 576;
export function contactDistances(players) {
  return players.map((p) =>
    players.map((q) =>
      p.team === q.team
        ? referenceDistance(q.x - p.x, q.z - p.z)
        : opponentDistance(p, q),
    ),
  );
}
// sub_D448/sub_D4AC: cancel movement toward each nearby standing opponent.
// Compensation precedes ordinary movement; it does not change velocity.
export function blockPlayerMovement(players, i, distances, dt) {
  const p = players[i];
  for (let j = 0; j < players.length; j++) {
    const q = players[j];
    if (q.team === p.team || q.stun > 0 || q.health <= 0 || distances[j] > 30)
      continue;
    if (p.action === 4 && p.stun <= fallRecovery + 1e-9) {
      p.fallX = p.fallZ = p.moveX = p.moveZ = 0;
      p.stun = p.actionTime = fallRecovery;
    }
    const dx = Math.round(q.x / unit) - Math.round(p.x / unit);
    const dz = Math.round(q.z / unit) - Math.round(p.z / unit);
    if (Math.abs(dx) > Math.floor(Math.abs(dz) / 2) && dx * p.moveX > 0)
      p.x -= p.moveX * dt;
    if (Math.abs(dz) > Math.floor(Math.abs(dx) / 2) && dz * p.moveZ > 0)
      p.z -= p.moveZ * dt;
  }
}

File: server/collision.go
package main

import "math"

func contactDistances(players *[18]Player) [18][18]int {
	var distances [18][18]int
	for i, p := range players {
		for j, q := range players {
			distances[i][j] = referenceDistance(q.X-p.X, q.Z-p.Z)
		}
	}
	return distances
}

// sub_D448/sub_D4AC: compensate before ordinary movement, retaining velocity.
func blockPlayerMovement(players *[18]Player, i int, distances *[18]int, dt float64) {
	p := &players[i]
	const unit = 22.4 / 576
	for j, q := range players {
		if q.Team == p.Team || q.Stun > 0 || q.Health <= 0 || distances[j] > 30 {
			continue
		}
		dx := int(math.Round(q.X/unit) - math.Round(p.X/unit))
		dz := int(math.Round(q.Z/unit) - math.Round(p.Z/unit))
		if absInt(dx) > absInt(dz)/2 && float64(dx)*p.moveX > 0 {
			p.X -= p.moveX * dt
		}
		if absInt(dz) > absInt(dx)/2 && float64(dz)*p.moveZ > 0 {
			p.Z -= p.moveZ * dt
		}
	}
}
func absInt(v int) int {
	if v < 0 {
		return -v
	}
	return v
}
